### PR TITLE
Asyncify without topmost interface

### PR DESCRIPTION
**Pull request by 2colours**
The second attempt, following https://github.com/pulsar-edit/ppm/pull/93

Long story short: I do think the approach was alright but given the test framework, basically it was a Russian roulette what tests could work, if at all.

This time around, I deliberately omitted the very part I started with last time so that the interface shown towards the executable scripts and the tests themselves would stay callback-based.

### CODE REVIEW

**Comment by 2colours** [comment]
So the first couple of functions could be changed no problem, without any test outage.

The big one was the wrapping of the `run` method of all commands - I isolated the failing tests because there are all sorts of weird interactions.

First, it would be great to fix these tests actually. I'm not sure if that's possible without altering the tests. As a final radical measure, I think these commands could get separate handling behind a wrapper so if all seems lost, maybe there is still a way to keep _only these commands_ callback-based, which I think is still a win, if the basic idea was that the code is ugly. The vast majority of code could still be not-so-ugly, and these commands could rightfully shout "think twice before touching me".

Then, for the other commands, I still think gradually going towards the bottom is a workable paradigm. There can be problems with the mocking but ultimately, if one part is fully Promise-based and eventually asyncified, there will be just genuinely no reason to try to mock certain functions by changing their arguments and such...

**Comment by 2colours** [comment]
I noticed a few things and refined the strategy as well.
- first, the commands that already work with the tests are processed
- then I think the other couple of commands need to be fixed by reducing the tests until a single point of failure remains, solving that, and then gradually adding the tests back if need be
- then lower layers of code can be changed; they will probably gradually allow the higher layers to be refined to a nicer code structure
- the interface of many Promises (what is a value, what is an error) will have to change. I started with a "whenever it's not crystal clear, use the resolve path only" policy. I think I have been overdoing it. Currently, I think only the Promises that actually have to pass resolve or reject on are okay to do that (if at all...). This will have to be redone when I reach back to the higher layers

Anyway, for Commands, the workflow is pretty simple:
1. check if some other command depends on directly calling this Command (with `run`). If it does, make sure it can be rewritten in a way that satisfies apm-cli and the other command as well, or else it may need to be skipped.
2. set `static promiseBased = true` for the class
3. go through the methods, checking their implementation to pick a promisification strategy
4. check the callers of the methods and rewrite those calls appropriately
5. as part of this process, rewrite the `run` method as well, check the tests, fix the mistakes...

**Comment by 2colours** [comment]
I reached a state where all tests succeed with slight alternations:
1. some functions need to be mocked differently - this is pretty obvious
2. the ultimate return value the callback obtains is sometimes downright undefined, not null - I think this is even desirable and there should be less explicit null returns but this is easy to twist in many ways

I would be worried about the `install --json` path; it's like a whole different command that is carried over to quite low layers of the same functions, and it is the reason why many installation methods need to produce values... I think it needs further testing at the very least; it would be nice to rework it once in the future.

**Comment by 2colours** [comment]
Now, this pull request has "officially" reached the state where (I think) everything that clearly needed an async-based implementation, got one.

The next - and arguably final - stage should consist of:
- "integration testing", i.e actually trying out various functionality, at least with a sample check approach
- refactoring of leftover stuff if we find such
- finalizing the interfaces that this version of code provides, so typically deciding on how the commands work
  - currently, their return value is the error, if there is any
  - and apm-cli did remain callback-based because otherwise all the tests should have been rewritten
  - I don't like these two things, especially the former... but I think we could actually live with them for the time being, and do different PR's to mitigate them

**Comment by 2colours** [comment]
From now on, there are no planned modifications that I still wanted to do, only fixes, or, if it's decided that some interface modification is required in this PR, then that can be added. All in all, I think it's ready for review.

As I said, since the spec is far from exhaustive, some testing is required, others are of course more than welcome to join in with that.

**Comment by DeeDeeG** [comment]
Rebasing on top of `master` branch now that #100 is merged, to basically port #100's changes to this PR and sync everything. As was discussed here on GitHub and on the Discord.

Old commit of this branch just before I rebased:  9dd95c6987c7b114e98ddb5a72356f55bb5ca47d
New commit after I rebased: 0e865d2e4f43c239fb1ff3e5f79ae88bc9bbb13a

File: src/install.js [hunk 1]
```diff
@@ -276,124 +280,120 @@ Run ppm -v after installing Git to see what version has been detected.\
...
-          if (!packageVersion) {
-            this.logFailure();
-            callback(`No available version compatible with the installed Atom version: ${this.installedAtomVersion}`);
-            return;
+      const commands = [];
+      try {
+        const pack = await this.requestPackage(packageName);
+        packageVersion ??= this.getLatestCompatibleVersion(pack);
+        if (!packageVersion) {
+          throw `No available version compatible with the installed Atom version: ${this.installedAtomVersion}`;
```
Comment (confused-Techie) [inline src/install.js:288]:
```suggestion
          throw `No available version compatible with the installed Pulsar version: ${this.installedAtomVersion}`;
```

Again not your error or fault, just a good idea to fix this while seeing it

**Review by confused-Techie — CHANGES_REQUESTED** [review]
Just wanna say this is some fantastic work! Seriously, huge undertaking and I'm very much impressed.

I've left a few comments around about improvements or questions, or even just some comments. So please address them when you've got the chance, but besides those few items overall this looks perfect, and as for the code side of this review, I'd say it's basically good to go! 

With that I'll add I'm relying on the manual testing that @DeeDeeG has done, and the automated tests to point out any unknown flaws in the logic, but as for the code it looks identical and completely valid, so I think we are good to go there!

File: src/apm-cli.js [hunk 2]
```diff
@@ -61,13 +61,13 @@ const commandClasses = [
...
 for (let commandClass of commandClasses) {
-  for (let name of commandClass.commandNames != null ? commandClass.commandNames : []) {
+  for (let name of commandClass.commandNames ?? []) {
     commands[name] = commandClass;
   }
 }
 
-const parseOptions = function(args) {
-  if (args == null) { args = []; }
+function parseOptions(args) {
```
[the remark this answers is not in the document]
Comment (2colours) [reply src/apm-cli.js:69]: The reason I didn't change it is that I think the functionality is _not_ the same. If these functions ever actually receive `null`, `args = []` in the argument list won't help to override it, only for `undefined`!

Of course one could go around and make sure `null` is never used to indicate a missing value but I think that falls way out of the scope of this pull request.

File: src/apm.js [hunk 3]
```diff
@@ -26,55 +25,57 @@ module.exports = {
...
-    let apmFolder = path.resolve(__dirname, '..');
-    let appFolder = path.dirname(apmFolder);
-    if ((path.basename(apmFolder) === 'ppm') && (path.basename(appFolder) === 'app')) {
-      asarPath = `${appFolder}.asar`;
-      if (fs.existsSync(asarPath)) {
-        return process.nextTick(() => callback(asarPath));
+  getResourcePath() {
+    return new Promise((resolve, _reject) => {
+      if (process.env.ATOM_RESOURCE_PATH) {
+        return void process.nextTick(() => resolve(process.env.ATOM_RESOURCE_PATH));
```
[the remark this answers is not in the document]
Comment (2colours) [reply src/apm.js:31]: There are two answers here, depending on what you compare to.

If you compare to `justMakeTheCall(); return;`, it's really the same thing, kind of stylistics. I tend to like the aesthetics and terseness of it, especially for arrow functions where they are "a clear win" regarding syntax. Some defined style guide could trump it at any second.

If you compare to `return justOneMoreCall();`, it is indeed "just" to ensure that we return `undefined` - but I don't find that so much of a "just" thing. A more clever static analyzer, or Typescript in particular, could give the users of a certain function type hints about the returned value. To let them know on all possible layers that the function is never ever going to produce any useful value, is a noble and sensible goal in my opinion. By applying this convention, it's immediately clear what statement is there to actually produce a value. This is kind of the opposite of the old Coffeescript way that deliberately didn't distinguish between returning values and just ending a function.

File: src/ci.js [hunk 4]
```diff
@@ -60,23 +60,27 @@ but cannot be used to install new packages or dependencies.\
...
-      if (err) { return callback(err); }
-      return callback(null);
+    commands.push(async () => {
+      const npm = await config.loadNpm();
+      this.npm = npm;
     });
+    commands.push(async () => await this.loadInstalledAtomMetadata());
+    commands.push(async () => this.installModules(opts));
+    try {
+      await async.waterfall(commands);
```
[the remark this answers is not in the document]
Comment (2colours) [reply src/ci.js:81]: I also felt that this is out of the scope of this PR, and honestly, this "waterfall" utility that pipes the values from function to the other didn't feel half bad in its final form.

File: src/command.js [hunk 5]
```diff
@@ -8,38 +8,34 @@ const git = require('./git');
 
 module.exports =
 class Command {
-  constructor() {
-    this.logCommandResults = this.logCommandResults.bind(this);
```
[the remark this answers is not in the document]
Comment (2colours) [reply src/command.js:12]: What rather surprised me that despite the elimination of this constructor, the descended classes still need to make a super() call to keep things working... I guess there is some sort of invisible constructor that sets the field on the final object... perhaps C++ does better in this regard.

The point of this abundance of bindings was that the methods were often passed around as callbacks. When you pass a function over (and in JS, methods are really all just normal functions!), you lose control over the `this` binding. Therefore I suppose it seemed safe to just "burn it into the function".

In the new code, methods are basically never passed on as a callback. If somebody wanted to do that anyway, perhaps it's more intuitive to wrap it into a lambda that preserves the method call in itself.

File: src/auth.js [hunk 6]
```diff
@@ -18,34 +18,32 @@ const tokenName = 'pulsar-edit.dev Package API Token';
...
-          return callback(null, token);
-        } else {
-          return Promise.reject();
-        }}).catch(function() {
-        let token;
-        if ((token = process.env.ATOM_ACCESS_TOKEN)) {
-          return callback(null, token);
-        } else {
-          return callback(`\
+  // returns the token as string or throws an exception
```
[the remark this answers is not in the document]
Comment (2colours) [reply src/auth.js:21]: Wouldn't be complete without that. :) The amount of documentation is still rather *under*whelming but hopefully the code itself turns more self-documenting...

File: src/link.js [hunk 7]
```diff
@@ -28,39 +28,33 @@ Run \`ppm links\` to view all the currently linked packages.\
...
+    async run(options) {
       options = this.parseOptions(options.commandArgs);
 
       const packagePath = options.argv._[0]?.toString() ?? ".";
       const linkPath = path.resolve(process.cwd(), packagePath);
 
       let packageName = options.argv.name;
       try {
-        if (!packageName) { packageName = CSON.readFileSync(CSON.resolve(path.join(linkPath, 'package'))).name; }
+        packageName ||= CSON.readFileSync(CSON.resolve(path.join(linkPath, 'package'))).name;
```
[the remark this answers is not in the document]
Comment (2colours) [reply src/link.js:39]: It's kind of defensive, honestly. When there are so many moving parts and unknown technologies, I wouldn't dare to say `name` cannot come out as the empty string and then have some special handling... When looking at the bigger picture, probably these parts can be improved.

File: src/apm-cli.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (confused-Techie) [reply src/apm-cli.js:69]: Ah very good catch. Didn't even consider the event of passing `null` to the function explicitly. So far enough, there are cases where it wouldn't be equal, and best not to modify it without that confidence

File: src/fs.js [hunk 8]
```diff
@@ -23,34 +23,36 @@ const fsAdditions = {
...
   },
 
-  mv(sourcePath, destinationPath, callback) {
-    return rm(destinationPath, function(error) {
-      if (error != null) {
-        return callback(error);
-      } else {
+  mv(sourcePath, destinationPath) {
+    return new Promise((resolve, reject) => {
+      rm(destinationPath, error => {
```
[the remark this answers is not in the document]
Comment (DeeDeeG) [reply src/fs.js:39]: Modern nodeJS has a `fs.rm` that's really nice, if I recall correctly, so we can soon drop the rimraf dependency entirely, I bet.

File: src/apm.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (DeeDeeG) [reply src/apm.js:31]: I personally prefer `justMakeTheCall(); return;`.

`return void justMakeTheCall();` causes some mental gymnastics for me, since `void` is so very uncommon in JavaScript.

It also feels overloaded to me: the `return` and `void justMakeTheCall();` are kind of unrelated, other than it shows the author's intend to to `justMakeTheCall()` as the last thing in the function body before exiting. I rather think `justMakeTheCall(); return;` is a much more conventional (and therefore readable) way to signal this in JavaScript.

Few people work on ppm, and I'm now up to speed on what `void` means in JS, so as one of few people working on this code I am not so opposed to it that I would insist on removing it. But part of me almost feels I should, in case anyone new ever hopes to contribute to ppm. I think conventional and usual-looking code would be more accessible to newcomers to the codebase.

**In summary: I wouldn't insist on changing to `doSomething(); return`, but I would suggest it be done nonetheless.** If it's not done in this PR and if I still feel strongly enough about it later, I can do a follow-up PR to tweak the code not to use `void`.

File: src/command.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (confused-Techie) [reply src/command.js:12]: Ahh that makes total sense @2colours, but now without all the passing around it then becomes pointless. So glad to see the change was thought out, was just confirming

File: src/ci.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (confused-Techie) [reply src/ci.js:81]: That's totally fair that it's out of scope. And I suppose it's not too bad. But it may be something I come back to eventually wanting to fix

File: src/fs.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (confused-Techie) [reply src/fs.js:39]: @DeeDeeG This is awesome news, guess there's another followup PR that we could tackle

File: src/apm.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (2colours) [reply src/apm.js:31]: I find the "commonness argument" fallacious because one kind of has to assume a practical reason behind something being uncommon, or else it's really just catch-22. I don't have a representative study down in my pocket but I don't think it's _that_ obscure anyway, it seems like the kind of thing you can read once and remember. Unfortunately, one can come across `return iOnlyReallyExecutedThisOneForTheSideEffects()` which is basically the semantically wrong version of it, and as I said, this is mostly an opt-out from that pattern.

Having said that, there could be extensions to a written or conventional style guide for code here. This isn't my project so I tried to preserve apparent conventions, even the ones I find just silly. For example, I quite vehemently oppose the famous `===` dogma (as it doesn't actually address typing problems, only confuses things further) or, what was even weirder in this code: using parens around basically all comparisons when used in combination with logic operators, as if it was Pascal. It's really not my business to either set guidelines or to challenge existing ones, I only did what felt like a clear win within the established conventions. Since the former status quo was to just return ad-hoc values even for practically void functions, I think eliminating that was a clear win even at the cost of having to learn one pretty old operator. :P

At the end of the day, though, while I wouldn't deliberately sabotage existing conventions, I don't want to put any effort into migrating my code to a previously non-existent convention I softly disagree with. It's probably pretty easy; I can imagine that there is a good enough regex to do 90% of that work but I think that's all the more reason to see who says "it's worth my time". I wouldn't say that, for one.

[the remark this answers is not in the document]
Comment (confused-Techie) [reply src/apm.js:31]: You do have some points about weird conventions that already exist here, even if they aren't our doing. But I suppose you have a point that it isn't the hardest to learn as new, 

Although at the end of the day, I suppose we leave this one be, since you obviously feel quite strong, and like it's been mentioned this isn't a blocker, so I'm willing to say we bank off that to see how soon we can get this in

**Comment by Spiker985** [comment]
@DeeDeeG @confused-Techie  Since you're more familiar with this PR, is this PR something that you believe we can get in before next release?

Looks like it may want a new review, confused

**Review by confused-Techie — APPROVED** [review]
Alright, I'm gonna add an approval on here, but do want to make a clarification.

I've reviewed only the code side of changes, ensuring accuracy, and coherence to how things previously functioned. Ensuring that the only essential changes were made, albeit with some tiny stylistic and quality of life improvements.

So the code itself looks good, my concerns have been addressed, and otherwise this PR looks incredibly solid.

But as for functionality, I'm trusting two things, one @DeeDeeG themselves having previously tested these changes, the author of this PR themselves also testing these changes, and lastly the fact that our specs are happy, and report running the exact same amount of tests as the most recent PRs here, indicating that not much has changed coverage wise (Hopefully).

With that said, I'll add my approval here, but would be more than happy to hear from @DeeDeeG about any of their concerns coverage or functionality wise, if they exist.

Otherwise I fully support us getting this in, and getting it updated in Pulsar itself

**Review by DeeDeeG — COMMENTED** [review]
I tested a bunch of stuff before, and the PR is frozen from any meaningful changes since then, so it still stands.

I may not have gotten to every subcommand, but I'm comfortable enough getting this into Rolling hopefully with a few days lead time before Regular. (Tangent: That timing would be kind of tight, and I don't want to rush people about it, to be honest. But this is a tad off-topic to this PR itself.)

If we would have to do a hotfix release of Pulsar to roll back `ppm` and do a "try two" of this PR (even if maintainers have to handle that) that is all doable.

So, at this point I can "lite approve" (meaning, despite best intentions, I didn't have time to review the code myself, so I am approving on a basis of being willing to troubleshoot and manage any issues post-merge.) Let's go for it, I think.

**Comment by DeeDeeG** [comment]
Thanks again @2colours for all the hard work on this! I am merging it now.

Hopefully some more `async`/`await` style code will make things a lot easier to read and deal with! Thanks!
